### DISCUSSION TITLE
Errors importing less and PNG/SVG in a React App

### DISCUSSION THREAD

**Discussion by cmaronchick** [babel/babel#11840] [Q&A] [answered]
I'm really struggling with this.

I would like to implement SSR to improve SEO for my app, and I am stuck.

I used CRA for the client side and express for the server.

webpack is successfully compiling the client, but when I run everything, I get errors related to my static files like less (I'm using Ant Design, FWIW) and SVG.

I feel like I have tried just about every possible permutation and nothing seems to be working, so I think I may be missing something fundamental.

My babel.config.js is:

```
module.exports = {
    "presets": ['@babel/preset-react', ['@babel/preset-env', {
        targets: {
          "browsers": ["last 2 versions"],
          node: "current"
        }
      }]],
    "plugins": [["@babel/plugin-proposal-class-properties"], ["@babel/plugin-transform-runtime", {
        "regenerator": true
      }]
}
```

My file structure is: /server/... /client/... and my npm commands are (I have tried both, and neither works):

```
"dev": "NODE_ENV=development webpack -wd & NODE_ENV=development nodemon --exec babel-node ./server/server.js --ignore public",
"server": "nodemon --exec babel-node ./server/server.js NODE_ENV=development --ignore public"
```

Here's an example of an error I receive:

I import a file with this: 
import './App.less';

which includes this line:
@import '~antd/dist/antd.less';

```
@import '~antd/dist/antd.less';
^

SyntaxError: Invalid or unexpected token
```

Thank you!

**Comment by JLHwung** [comment, marked as answer]
Please install `less-loader`: https://webpack.js.org/loaders/less-loader/

Babel does not process `.less` stylesheets.

**Comment by cmaronchick** [comment]
Thank you. I have less-loader installed and working with webpack, so I must be doing something wrong with babel.

This error occurs with SVGs and PNGs as well.

I must be having babel-node trying to parse files that haven't been compiled, but I don't know what I need to change in order to fix it.

**Reply by JLHwung** [reply]
```
@import '~antd/dist/antd.less';

SyntaxError: Invalid or unexpected token
```

Can you attach the error stack? Where is this error is thrown from?

It will be really helpful if you can share a reproducing repo. Otherwise I don't have enough information to tell what is going wrong.

**Reply by cmaronchick** [reply]
I'd be glad to, thank you.

```
/Users/chrisaronchick/GitHub/crowd-wisdom-sports-react-v2/client/src/App.less:1
@import '~antd/dist/antd.less';
^

SyntaxError: Invalid or unexpected token
    at wrapSafe (internal/modules/cjs/loader.js:1047:16)
    at Module._compile (internal/modules/cjs/loader.js:1097:27)
    at Module._extensions..js (internal/modules/cjs/loader.js:1153:10)
    at Object.newLoader [as .js] (/Users/chrisaronchick/GitHub/crowd-wisdom-sports-react-v2/node_modules/pirates/lib/index.js:104:7)
    at Module.load (internal/modules/cjs/loader.js:977:32)
    at Function.Module._load (internal/modules/cjs/loader.js:877:14)
    at Module.require (internal/modules/cjs/loader.js:1019:19)
    at require (internal/modules/cjs/helpers.js:77:18)
    at Object.<anonymous> (/Users/chrisaronchick/GitHub/crowd-wisdom-sports-react-v2/client/src/App.js:9:1)
    at Module._compile (internal/modules/cjs/loader.js:1133:30)
```

from my serverRender file:

```
import React from 'react';
import ReactDOMServer from 'react-dom/server';
import { StaticRouter } from 'react-router-dom';

import App from '../client/src/App';
```

Here's my webpack file:

```
const path = require('path');

module.exports = {
  entry: ['@babel/polyfill','./client/src/app.js'],
  output: {
    path: path.resolve('public'),
    filename: 'bundle.js'
  },
  module: {
    rules: [
      {
        test: /\.js$/,
        exclude: /node_modules/,
        use: {
          loader: 'babel-loader',
          options: {
            presets: [
                "@babel/preset-env",
                "@babel/preset-react"
            ]
          }
        },
      },
      {
        test: /\.less$/, // match any .less file, 
        use: [
          {loader: 'style-loader'},
          {loader: 'css-loader'},
          { loader: "less-loader",
            options: {
              lessOptions: {
                modifyVars: { 
                  '@primary-color': '#0a1f8f',
                  '@secondary-color': '#124734',
                  '@tertiary-color': '#e04403',
                  '@info-color': '@primary-color',
                  '@success-color': '#124734',
                  '@processing-color': '@primary-color',
                  '@error-color': 'rgba(224, 20, 7, 1.0)',
                  '@highlight-color': '#f6dfa4',
                  '@warning-color': 'rgba(224, 20, 7, 1.0)',
                  '@normal-color': '#231f20',
                  '@white': '#fff',
                  '@black': '#231f20',
                  '@layout-header-background': '#231f20',
                  '@layout-header-color': '#f6dfa4'
                  },
                  javascriptEnabled: true
              }
            }
          }
        ]
      },
      
      {
        test: /\.css$/, // match any .scss or .css file, 
        use: [
          { loader: "style-loader" }, 
          { loader: "css-loader" },
          
        ]
      },
      {
        test: /\.(png|jpe?g|gif|svg)$/i,
        use: [
          {
            loader: 'file-loader',
          },
        ],
      }
    ]
  }
};
```

Thank you for your help. I'll be glad to provide any other information you may need.

**Reply by cmaronchick** [reply]
Just adding one more thing that may help:
I added the "import-static-files" plugin to my babel.config.js file and I no longer received the PNG error, so I have to believe that my babel-node call is not picking up the output of webpack, but again, I have no idea why it isn't and what to fix.

**Reply by JLHwung** [reply]
The error is thrown from node.js, which means node is loading the `.less` files and this is not expected.

I believe for SSR, the `style-loader` will not work as intended -- we don't have DOM on servers.

You may consider replace `style-loader` by https://github.com/kriasoft/isomorphic-style-loader which claims for SSR supports.

I don't think Babel is involved here -- Babel does _not_ process your imports -- they are handled by webpack and webpack will call babel-loader, based on the module file extension. If a `.less` file is passed to `babel-loader`, the syntax error will be thrown from Babel.

**Reply by jayliew** [reply]
@cmaronchick I have the same issue :( Did you find a solution after all?
